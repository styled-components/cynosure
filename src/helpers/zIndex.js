// @flow

/**
 * Returns the z-index value in array or object
 * first parameter: key in the array or object
 * second paramater: z-index array or object
 *   if it is array, first element of array
 *   will have the highest z-index value, last element of array will have
 *   the lowest z-index value
 *   if it is object, it will return the z-index of key
 * third parameter (optional): base z-index. this helper function will add the base value
 *   to the z-index value of the key.
 *
 * @example
 * const zArray = ['modal', 'backDrop', 'lightbox', 'header', 'base']
 * // Styles as object usage
 * const styles = {
 *   ...zIndex('header', zArray)
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${zIndex('header', zArray)}
 * `
 *
 * // CSS in JS Output
 *
 * div {
 *   z-index: 2;
 * }
 */

function zIndex(layer: string, layers: Array | Object, base: number = 0) {
  if (!layer) throw new Error('zIndex expects a layer name in first parameter')

  if (Array.isArray(layers)) {
    const reversedLayers = layers.slice().reverse()
    return {
      'z-index': base + reversedLayers.indexOf(layer),
    }
  } else if (typeof layers === 'object') {
    return {
      'z-index': layers[`${layer}`],
    }
  } else {
    throw new Error('zIndex expects an array or object of layers in second parameter')
  }
}

export default zIndex
