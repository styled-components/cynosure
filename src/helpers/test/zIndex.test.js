// @flow
import zIndex from '../zIndex'

describe('stripUnit', () => {
  const layersArray = [
    'modal',
    'backDrop',
    'header',
    'goTop',
    'base',
  ]
  const layersObject = {
    header: 10,
    modal: 23,
    goTop: 6,
    base: 1,
    backDrop: 18,
  }

  it('should return correct z-index value in array', () => {
    expect({ ...zIndex('modal', layersArray) }).toMatchSnapshot()
  })

  it('should return correct z-index value in array with base value', () => {
    expect({ ...zIndex('modal', layersArray, 100) }).toMatchSnapshot()
  })

  it('should return correct z-index value in object', () => {
    expect({ ...zIndex('modal', layersObject) }).toMatchSnapshot()
  })

  it('should not allow when any parameter is not defined', () => {
    expect(() => {
      zIndex()
    }).toThrow()
  })

  it('should not allow when layers array or object parameter is not defined', () => {
    expect(() => {
      zIndex('modal')
    }).toThrow()
  })

  it('should properly add rules when block has existing rules', () => {
    expect({
      background: 'red',
      ...zIndex('base', layersArray),
    }).toMatchSnapshot()
  })
})
