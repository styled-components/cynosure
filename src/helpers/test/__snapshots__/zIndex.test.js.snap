exports[`stripUnit should properly add rules when block has existing rules 1`] = `
Object {
  "background": "red",
  "z-index": 0,
}
`;

exports[`stripUnit should return correct z-index value in array 1`] = `
Object {
  "z-index": 4,
}
`;

exports[`stripUnit should return correct z-index value in array with base value 1`] = `
Object {
  "z-index": 104,
}
`;

exports[`stripUnit should return correct z-index value in object 1`] = `
Object {
  "z-index": 23,
}
`;
